community/kernel/src/main/java/org/neo4j/kernel/internal/DatabaseHealth.java

```diff
@@ -88,21 +89,35 @@ public void panic( Throwable cause )
             throw new IllegalArgumentException( "Must provide a cause for the database panic" );
         }
         this.causeOfPanic = cause;
-        this.tmOk = false;
+        this.healthy = false;
         log.error( "Database panic: " + panicMessage, cause );
         dbpe.generateEvent( ErrorState.TX_MANAGER_NOT_OK, causeOfPanic );
     }
 
     public boolean isHealthy()
     {
-        return tmOk;
+        return healthy;
     }
 
-    public void healed()
+    public boolean healed()
     {
-        tmOk = true;
-        causeOfPanic = null;
-        log.info( "Database health set to OK" );
+        if ( hasCriticalFailure() )
+        {
+            log.error( "Database encountered a critical error and can't be healed. Restart required." );
+            return false;
+        }
+        else
+        {
+            healthy = true;
+            causeOfPanic = null;
+            log.info( "Database health set to OK" );
+            return true;
+        }
+    }
+
+    private boolean hasCriticalFailure()
+    {
+        return !isHealthy() && Exceptions.contains( causeOfPanic, CRITICAL_EXCEPTIONS );
```

Comment:
`hasCriticalFailure` will only report failure to be critical if first `cause` when `panic` was called was critical. Else, if some other `cause` that IS critical is later reported (before `healed`), then this will be ignored.